tests/h/services/bulk_executor/conftest.py

```diff
@@ -1,36 +1,58 @@
 from h.h_api.bulk_api import CommandBuilder
 
-AUTHORITY = "lms.hypothes.is"
 
-
-def make_group_command(authority=AUTHORITY, query_authority=AUTHORITY):
-    command = CommandBuilder.group.upsert(
-        {
-            "name": "name",
-            "authority": query_authority,
-            "authority_provided_id": "authority_provided_id",
-        },
-        "id_ref",
-    )
-
-    # Fake the effect of merging in the query
-    command.body.attributes["authority"] = authority
-
-    return command
-
-
-def make_user_commmand(authority=AUTHORITY, query_authority=AUTHORITY):
-    command = CommandBuilder.user.upsert(
-        {
-            "username": "username",
-            "display_name": "display_name",
-            "authority": query_authority,
-            "identities": [{"provider": "p", "provider_unique_id": "pid"}],
-        },
-        "id_ref",
-    )
-
-    # Fake the effect of merging in the query
-    command.body.attributes["authority"] = authority
-
-    return command
\ No newline at end of file
+class CommandFactory:
+    """Create commands for the bulk tests."""
```

Comment:
I think this class belongs in [`tests/common/factories`](https://github.com/hypothesis/h/tree/master/tests/common/factories) rather than in `conftest.py`, that's where we keep all our test factories and they're automatically made available to tests via the `factories` fixture.

It should probably be using [factoryboy](https://factoryboy.readthedocs.io/en/latest/) as well, as that's what our other factories use, and it automates what this class is implementing for itself in a really handy way. For example it has builtin support for autoincrementing sequence numbers for `username="user_1"` and `display_name="display_name_1"` etc. It also automatically supports passing in extra values to override or extend the ones that would normally be generated.

The code would look something like below. This would have to be modified to work with how the bulk_api models actually are, but you get the idea. You write one factory class per model (not a single `CommandFactory` with multiple classmethods). Below's an example of a `UserUpsertCommand` factory.

Usage would be something like:

```python
def test_foo(factories):
    # An UpsertUserCommand model with username=user_1, display_name=display_name_1, etc.
    cmd1 = factories.UpsertUserCommand()

    # This one will get username=user_2, display_name=display_name_2, etc.
    # The sequence automatically increments, and then resets to 1 for the next test.
    cmd2 = factories.UpsertUserCommand()

    # Pass in an arbitrary extra field name/value and it will be passed
    # to the model class's __init__.
    cmd3 = factories.UpsertUserCommand(foo="bar")

    # Or override any of the default fields while letting the rest still be autogenerated.
    cmd4 = factories.UpsertUserCommand(username="foo", display_name="bar")
```

The code to implement the factory (not based on the actual bulk_api models, just a simple example):

```python
import factory

from bulk_api import models


class UpsertUserCommand(factory.Factory):
    class Meta:
        model = models.UpsertUserCommand

    username = factory.Sequence(lambda n: f"user_{n}")
    display_name = factory.Sequence(lambda n: f"display_name_{n}")
    authority = "lms.hypothes.is"
    identities = factory.Sequence(
        lambda n: [{"provider": "provider", "provider_unique_id": f"pid_{n}"}]
    )
```

Reply:
I've moved the code out to factories, but we still have to import it as it's used to create parameterised fixtures. I've also just used basic dict merging and kwargs instead of a separate function.

I don't think `factoryboy` is going to be a good fit here for reasons being discussed in prod-general. I think we'd have to subvert or avoid most of what `factoryboy` does to get this to work. but I'd be interested in seeing a working example.

Reply:
As per our discussion:

* I've moved this back out of factories and renamed the file `factories.py`
* The calls are now functions with "object-like" names to fit the interface of factories elsewhere. It's a function, but you still call it like `UpserUserCommand(blah...)`

Reply:
A `factory_boy` based implementation has been added for comparison. Both work in the tests, but we should pick one and remove the other before merging.

I think it might be good to take a quick straw poll and then see if people prefer A or B and go with that.